esp-hal/src/dma/mod.rs

```diff
@@ -1657,6 +1657,91 @@ impl<DEG: DmaChannel> DmaChannelConvert<DEG> for DEG {
     }
 }
 
+/// Trait implemented for DMA channels that are compatible with a particular
+/// peripheral.
+///
+/// You can use this in places where a peripheral driver would expect a
+/// `DmaChannel` implementation.
+#[cfg_attr(pdma, doc = "")]
+#[cfg_attr(
+    pdma,
+    doc = "Note that using mismatching channels (e.g. trying to use `spi2channel` with SPI3) may compile, but will panic in runtime."
+)]
+#[cfg_attr(pdma, doc = "")]
+/// ## Example
+///
+/// The following example demonstrates how this trait can be used to only accept
+/// types compatible with a specific peripheral.
+///
+/// ```rust,no_run
+#[doc = crate::before_snippet!()]
+/// use esp_hal::spi::master::{Spi, SpiDma, Config, Instance as SpiInstance};
+/// use esp_hal::dma::CompatibleWith;
+/// use esp_hal::peripheral::Peripheral;
+/// use esp_hal::Blocking;
+/// use esp_hal::dma::Dma;
+///
+/// fn configures_spi_dma<'d, S, CH>(
+///     spi: Spi<'d, Blocking, S>,
+///     channel: impl Peripheral<P = CH> + 'd,
+/// ) -> SpiDma<'d, Blocking, S>
+/// where
+///     S: SpiInstance,
+///     CH: CompatibleWith<S> + 'd,
+///  {
+///     spi.with_dma(channel)
+/// }
+///
+/// let dma = Dma::new(peripherals.DMA);
+#[cfg_attr(pdma, doc = "let dma_channel = dma.spi2channel;")]
+#[cfg_attr(gdma, doc = "let dma_channel = dma.channel0;")]
+#[doc = ""]
+/// let spi = Spi::new_with_config(
+///     peripherals.SPI2,
+///     Config::default(),
+/// );
+///
+/// let spi_dma = configures_spi_dma(spi, dma_channel);
+/// # }
+/// ```
+pub trait CompatibleWith<P: DmaEligible>: DmaChannel + DmaChannelConvert<DmaChannelFor<P>> {}
```

Comment:
Just to throw my vote in name-wise. I think `DmaChannelFor` is a better name than `CompatibleWith` for what users will be writing.

I haven't look into the consequences of making this happen :smile:, was just my two cents on the public name.

2 cents aside.

Are these still needed now? Seems like users won't be using it directly anymore.
```Rust
/// Helper type to get the DMA (Rx and Tx) channel for a peripheral.
pub type DmaChannelFor<T> = <T as DmaEligible>::Dma;
/// Helper type to get the DMA Rx channel for a peripheral.
pub type RxChannelFor<T> = <DmaChannelFor<T> as DmaChannel>::Rx;
/// Helper type to get the DMA Tx channel for a peripheral.
pub type TxChannelFor<T> = <DmaChannelFor<T> as DmaChannel>::Tx;
```

Reply:
> Seems like users won't be using it directly anymore.

Not the users, but we are, and they really help :D But this naming clashes with the `CompatibleWith` types but they do different things 🤔 I don't really know what would be best.

Reply:
> Not the users, but we are, and they really help

Ah true, #[docs(hidden)] then? Rust might actually allow them to be `pub(crate)` but not sure.

I'll leave everyone else to pitch in on the names.

Reply:
I think I can merge them into Channel so that we'd have `Channel<'d, DmaMode, Peripheral>`

Reply:
Certainly a lot better than `ChannelCreator<2>: DmaChannelConvert<<T as DmaEligible>::Dma>,` 🙃